crates/polars-plan/src/client/mod.rs

```diff
@@ -1,68 +1,18 @@
-mod dsl;
+mod check;
+mod sink;
 
 use polars_core::error::{polars_err, PolarsResult};
-use polars_io::path_utils::is_cloud_url;
 
-use crate::dsl::Expr;
-use crate::plans::options::SinkType;
-use crate::plans::{DslFunction, DslPlan, FunctionNode};
+use crate::plans::DslPlan;
 
-/// Assert that the given [`DslPlan`] is eligible to be executed on Polars Cloud.
-pub fn assert_cloud_eligible(dsl: &DslPlan) -> PolarsResult<()> {
-    let mut expr_stack = vec![];
-    for plan_node in dsl.into_iter() {
-        match plan_node {
-            DslPlan::MapFunction {
-                function: DslFunction::FunctionNode(function),
-                ..
-            } => match function {
-                FunctionNode::Opaque { .. } => return ineligible_error("contains opaque function"),
-                #[cfg(feature = "python")]
-                FunctionNode::OpaquePython { .. } => {
-                    return ineligible_error("contains Python function")
-                },
-                _ => (),
-            },
-            #[cfg(feature = "python")]
-            DslPlan::PythonScan { .. } => return ineligible_error("contains Python scan"),
-            DslPlan::GroupBy { apply: Some(_), .. } => {
-                return ineligible_error("contains Python function in group by operation")
-            },
-            DslPlan::Scan { paths, .. }
-                if paths.lock().unwrap().0.iter().any(|p| !is_cloud_url(p)) =>
-            {
-                return ineligible_error("contains scan of local file system")
-            },
-            DslPlan::Sink { payload, .. } => {
-                if !matches!(payload, SinkType::Cloud { .. }) {
-                    return ineligible_error("contains sink to non-cloud location");
-                }
-            },
-            plan => {
-                plan.get_expr(&mut expr_stack);
+/// Prepare the given [`DslPlan`] for execution on Polars Cloud.
+pub fn prepare_cloud_plan(dsl: DslPlan, uri: String) -> PolarsResult<Vec<u8>> {
+    let dsl = sink::add_sink(dsl, uri);
```

Comment:
- nit: I would swap these as that saves an extra iteration /check.
- For now we should maybe check if the `uri` is an s3 link.

nit: Does it also require a separate module and function? As I would expect this to be only called once and is 5ish loc.

Reply:
> nit: I would swap these as that saves an extra iteration /check.
> For now we should maybe check if the uri is an s3 link.

Right - I somehow had in mind that the DSL check would also verify the URI but it doesn't. Will update 👍 

> nit: Does it also require a separate module and function?

Not really - I was a bit on the fence about this one. I just generally try to avoid putting much logic in `mod.rs` files. And I would expect a lot more logic to be added here when we expand this logic (e.g. support other file types). Let me know if you want me to move this up into the `mod.rs`.

Reply:
I see them as orthogonal:

- For the function: I'd rather see it factored out when it truly has multiple callers, or does a significant amount of work.

- For the module: I agree with the sentiment about `mod.rs`, but I would wait a bit and factor out the `sink` related work once there is a bit more. Given that I think the function should not even exist, I think the module is definitely a bit much if we remove that function. :smile: